Project-Descriptions-and-Plans/CV32E4_TBD/Project_Concept_for_CV32E4_TBD_Nov-27_2023.md

```diff
@@ -0,0 +1,250 @@
+# OpenHW Project Concept Proposal: CV32E4_TBD
+*The PC proposal introduces the project and explains the market drivers and the "why"*
+
+# Title of Project - "CV32E4_TBD"
+# Project Concept Proposal
+## Date of proposal - 2023-11-27
+## Author(s) - Manuel PEZZIN (CEA)
+
+## High Level Summary of project, project components, and deliverables
+This proposal is a continuation of CV32E40Pv2 project.
+Main goal is to support official RVB (and a part of RVK) and (upcoming) RVP
+extensions while keeping backward compatibility with CV32E40Pv1/v2 (eg XPULP
+custom extensions are kept and even extended).
+Additionnal features may also be considered, like:
+* CV-X-IF support
+* CLIC support
+* Improved OBI interface support (bus error in particular)
+
+This new project will keep functionnal backward compatibility with CV32E40Pv2,
+but with relaxed rules compared to previous project. In particular, formal equivalence
+with previous design will not be required to make design work easier.
+
+Due to the strict backward compatibility rules of CV32E40Pv2, this new core is not
+named CV32E40Pv3 and a new naming should be found.
+
+## Summary of market or input requirements
+
+
+### Known market/project requirements at PC gate
+
+**Madatory features and design changes**
+
+Keep XPULP custom extensions that improve address manipulations and flow control:
+* Post-increment and register-register indexed load/store
+* Hardware loops
+* Immediate branch
+* Event load
+
+Introduce a dual opcode scheme for XPULP ALU instructions that have a strict RVB/RVP equivalent:
+* Multiply-accumulate
+* General ALU extensions
+* Bit manipulation
+* SIMD (16 and 8-bit data)
+note : this dual opcode scheme allows backard compatibility and open the door to smooth transition
+of existing programs from XPULP to RVB/RVP, if needed/relevant.
+
+Add support for RISC-V official bit manipulation instructions:
+* RVB zba (address manipulation) will be supported
+* RVB zbb (basic bit manipulations) will be supported
+* RVB zbs (single bit instructions) will be supported
+* RVB zbc (carry-less multiplications) : may be supported, to be confirmed if relevant for DSP applications
+* RVK zbkb (Bitmanip instructions for Cryptography) : pack, packh, packw, brev8, zip, unzip additionnal instructions may be supported, if relevant for DSP applications.
+* RVK zbkx (Crossbar permutation instructions) : xpem4 and xperm8 additionnal instructions may be supported, if relevant for DSP applications.
+* RVK zkn, zks, zkr and zkt are specific to cryptographic computations and will not be supported
+
+**Optionnal design changes**
+
+Introduce XPULPv3 custom extentions (preliminary, feature list may evolve):
+* ALU instructions that have a strict RVB/RVP equivalent opcode will be deprecated, new XPULPv3 opcodes will be RVB/RVP ones instead
```

Comment:
Here it is not clear - previously we wrote that we keep binary compatibility, here we say that we deprecate some of those - is it something you want to do in the future, with a parameter? and if so, how do you verify the CPU with xpulp and xpulpv3 extensions? do you freeze both versions in 2 different stages?

Reply:
Advertising a feature as deprecated doesn't mean it is unmaintained or removed in the current version.
This is more to inform for the end user that this feature will be removed in a future version and that SW should be updated to use new equivalent features if a long term product plan (using future version of this core) is envisioned.

Reply:
Regarding XPULPv3, maintaining dual instruction encoding for the first version will allow reusing current XPULP verification for the first version of the core. Then, the new instruction encoding and new instructions will be managed as an add-on to existing test suite.
Deprecated encodings will be removed from test suite of a future revision of the core, when actual removal is decided.

Reply:
but will the XPULPv3 be part of this core or not? and if so, with a parameter? what do you plan to verify as first?

Reply:
Yes XPULPv3 will be part of this core. What I have in mind is to extend the meaning of COREV_PULP parameter to play with features granularity. For example this could be:
- 0 : no XPULP feature at all
- 1 : full XPULPv2 support
- 2 : XPULPv3 support only
- 3 : XPULPv2 + XPULPv3 support
Another option could be a split of XPULP extensions into subset that have a corresponding gcc flag (HW loop, load/stores, SIMD, bitmanip, complex...) and add new flags for v3 support.

In any case, what I plan to verify first is the full XPULPv2 support.